slam-libraries/viam-cartographer/src/slam_service/slam_service_test.cc

```diff
@@ -0,0 +1,132 @@
+#include "slam_service.h"
+
+#include <boost/test/unit_test.hpp>
+#include <exception>
+
+namespace viam {
+namespace {
+
+BOOST_AUTO_TEST_SUITE(SLAMService)
+
+void checkCartoMapBuilderParameters(SLAMServiceImpl& slamService) {
+    auto pose_graph_options =
+        slamService.mapBuilder.map_builder_options_.pose_graph_options();
+    auto overlapping_submaps_trimmer_2d =
+        pose_graph_options.overlapping_submaps_trimmer_2d();
+    auto ceres_scan_matcher_options =
+        pose_graph_options.constraint_builder_options()
+            .ceres_scan_matcher_options();
+    auto trajectory_builder_2d_options =
+        slamService.mapBuilder.trajectory_builder_options_
+            .trajectory_builder_2d_options();
+
+    auto tolerance = boost::test_tools::tolerance(0.00001);
+    BOOST_TEST(pose_graph_options.optimize_every_n_nodes() ==
+               slamService.optimize_every_n_nodes);
+    BOOST_TEST(
+        trajectory_builder_2d_options.submaps_options().num_range_data() ==
+        slamService.num_range_data);
+    BOOST_TEST(trajectory_builder_2d_options.missing_data_ray_length() ==
+                   slamService.missing_data_ray_length,
+               tolerance);
+    BOOST_TEST(
+        trajectory_builder_2d_options.max_range() == slamService.max_range,
+        tolerance);
+    BOOST_TEST(
+        trajectory_builder_2d_options.min_range() == slamService.min_range,
+        tolerance);
+    if (slamService.GetActionMode() == SLAMServiceActionMode::LOCALIZING) {
+        BOOST_TEST(slamService.mapBuilder.trajectory_builder_options_
+                       .pure_localization_trimmer()
+                       .max_submaps_to_keep() ==
+                   slamService.max_submaps_to_keep);
+    } else {
+        BOOST_TEST(slamService.mapBuilder.trajectory_builder_options_
+                       .pure_localization_trimmer()
+                       .max_submaps_to_keep() == 0);
+    }
+    if (slamService.GetActionMode() == SLAMServiceActionMode::UPDATING) {
+        BOOST_TEST(overlapping_submaps_trimmer_2d.fresh_submaps_count() ==
+                   slamService.fresh_submaps_count);
+        BOOST_TEST(overlapping_submaps_trimmer_2d.min_covered_area() ==
+                       slamService.min_covered_area,
+                   tolerance);
+        BOOST_TEST(overlapping_submaps_trimmer_2d.min_added_submaps_count() ==
+                   slamService.min_added_submaps_count);
+    } else {
+        BOOST_TEST(overlapping_submaps_trimmer_2d.fresh_submaps_count() == 0);
+        BOOST_TEST(overlapping_submaps_trimmer_2d.min_covered_area() == 0,
+                   tolerance);
+        BOOST_TEST(overlapping_submaps_trimmer_2d.min_added_submaps_count() ==
+                   0);
+    }
```

Comment:
This is a bit tricky - when running in LOCALIZING or UPDATING mode, the internal cartographer parameters are changed to what the user defines, and what is saved in the slamService object for those values. But when one of these action modes is not activated, the internal parameters are set to 0.